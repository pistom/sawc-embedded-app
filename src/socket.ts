import { Manager } from 'socket.io-client';

const manager = new Manager(`http://${window.location.hostname}:3001`);

const socket = manager.socket("/", {
  auth: {
    token: window.localStorage.getItem("token"),
  }
});

export default socket;