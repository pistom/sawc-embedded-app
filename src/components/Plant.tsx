import { useContext, useEffect, useState } from "react";
import socket from '../socket.ts';
import Timer from "./Timer";
import { Cog6ToothIcon, LockOpenIcon } from '@heroicons/react/24/solid'
import { Link } from "react-router-dom";
import './plant.css';
import { UserMessagesContext } from "../context/UserMessagesContext";

interface PlantProps {
  device: DeviceConfig;
  output: OutputConfig;
  plantMessage?: Message;
  remainingTime?: { wateringTime: number, wateringIn: number, wateringVolume: number };
}

export default function Plant({ device, output, plantMessage, remainingTime }: PlantProps) {
  const { id: deviceId, settings: deviceSettings } = device;
  const { id, name, image, defaultVolume, sync } = output;
  const [isOn, setIsOn] = useState<boolean>(false);
  const [isWatering, setIsWatering] = useState<boolean>(false);
  const [wateringVolume, setWateringVolume] = useState<number>(defaultVolume || deviceSettings.defaultVolume);
  const [wateringVolumeRaw, setWateringVolumeRaw] = useState<number | string>(defaultVolume || deviceSettings.defaultVolume);
  const [initialWateringTime, setInitialWateringTime] = useState<number>(0);
  const [remainingWateringTime, setRemainingWateringTime] = useState<number>(0);
  const [wateringIn, setWateringIn] = useState<number>(0);
  const [isDisabledWateringBtn, setIsDisabledWateringBtn] = useState<boolean>(false);
  const userMessagesContext = useContext(UserMessagesContext);
  const { addMessage } = userMessagesContext;

  useEffect(() => {
    if (remainingTime) {
      setWateringVolume(remainingTime.wateringVolume);
      if (remainingTime.wateringIn < 0) {
        setIsWatering(true);
        setRemainingWateringTime(remainingTime.wateringTime + remainingTime.wateringIn);
        setInitialWateringTime(remainingTime.wateringTime);
      } else {
        setWateringIn(() => remainingTime.wateringIn);
        setWateringVolume(remainingTime.wateringVolume);
        setInitialWateringTime((initialWateringTime) => initialWateringTime === 0 ? remainingTime.wateringTime : initialWateringTime);
      }
      setIsOn(true);
    }
  }, [remainingTime]);

  useEffect(() => {
    if (plantMessage) {
      switch (plantMessage.status) {
        case "done":
        case "error":
        case "stopError":
          setTimeout(() => {
            setIsOn(false);
            setIsWatering(false);
            if (plantMessage.status === "error") {
              switch (plantMessage.context?.errno) {
                case -113:
                case "EHOSTUNREACH":
                  addMessage({
                    type: 'error',
                    title: `Can not reach network module (${plantMessage.context?.address})`,
                    message: 'Check that the device is switched on and properly configured.'
                  });
                  break;
                case -123:
                  addMessage({
                    type: 'error',
                    title: `Output number not found (${plantMessage.device}, output: ${plantMessage.output})`,
                    message: 'Check the device configuration.'
                  });
                  break;
                case "INVALIDTOKEN":
                  addMessage({
                    type: 'error',
                    title: `Invalid token`,
                    message: 'Check the device configuration.'
                  });
                  break;
              }
            }
          }, 1000);
          setWateringIn(0);
          break;
        case "aborted":
        case "stopped":
          setIsOn(false);
          setWateringIn(0);
          setIsWatering(false);
          break;
        case "watering":
          setIsOn(true);
          setIsWatering(true);
          setInitialWateringTime((initialWateringTime) => initialWateringTime === 0 ? plantMessage.duration : initialWateringTime);
          setRemainingWateringTime(plantMessage.duration);
          setWateringIn(0);
          break;
        case "calibratingError":
          console.error('calibratingError', plantMessage);
          break;
      }
    }
  }, [plantMessage, addMessage]);

  const handleWateringVolume = (e: React.ChangeEvent<HTMLInputElement>) => {
    setWateringVolumeRaw(e.target.value);
    if (e.target.value !== "" && parseInt(e.target.value) > 0) {
      setIsDisabledWateringBtn(false);
      let value = parseInt(e.target.value);
      if (value > deviceSettings.maxVolumePerOutput) {
        value = deviceSettings.maxVolumePerOutput;
      }
      setWateringVolume(value);
      setInitialWateringTime(0);
    } else {
      setIsDisabledWateringBtn(true);
    }
  }

  const handleMessageSubmit = (id: string, volume: number) => {
    if (!isOn) {
      socket && socket.emit("message", { action: 'startWater', device: deviceId, output: id, volume });
    } else {
      socket && socket.emit("message", { action: 'stopWater', device: deviceId, output: id });
    }
  };
  
  let imageSrc = image ? `/plants/${image}` : "/plant.svg";
  !name && (imageSrc = "/no-plant.svg");

  return <div className="w-full md:w-1/2 lg:w-1/3">
    <div className="plant relative">
      {!sync && !isOn && <span className="absolute bottom-2 right-2 text-sm text-slate-400 ">Not synced <LockOpenIcon className="align-sub h-5 w-5 inline-block" /></span>}
      <img className="image bg-slate-200" src={imageSrc} alt="" />
      <div className="details">
        {!isOn &&
          <Link to={`/output/edit/${deviceId}/${id}`} className="editBtn">
            <Cog6ToothIcon className="h-5 w-5 text-slate-300 absolute top-2 right-2" />
          </Link>}
        <h5 className="title">{name ?? `Output ${id}`}</h5>
        <div className="form flex gap-3 mb-3 font-normal text-gray-700">
          <div className="flex items-center">
            <input className="unit" type="number" disabled={isWatering || wateringIn > 0} value={wateringVolumeRaw} onChange={handleWateringVolume} />
            <span className="unit-label">ml</span>
          </div>
          <button className={`waterBtn btn ${!isOn ? `bg-emerald-700` : `bg-slate-700`}`} disabled={isDisabledWateringBtn} onClick={() => handleMessageSubmit(id, wateringVolume)}>
            {!isOn ? `Water` : wateringIn > 0 ? `Abort` : `Stop`}
          </button>
        </div>
        {wateringIn > 0 && (<Timer type="scheduled" duration={wateringIn} />)}
        {isWatering && <Timer type="current" duration={remainingWateringTime} initial={initialWateringTime} />}
      </div>
    </div>
  </div>

}