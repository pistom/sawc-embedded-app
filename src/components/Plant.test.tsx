import  socket  from '../socket.ts';
jest.mock('../socket.ts', () => {
  return {
    __esModule: true,
    default: {on: jest.fn(() => 42)},
  };
});
import { render } from "@testing-library/react";
import Plant from "./Plant";

describe('<Plant></Plant>', () => {
  it('should render the component', () => {
    const { container } = render(<Plant device="device" output="output" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should call socket on message when component is rendered', () => {
    render(<Plant device="device" output="output" />);
    expect(socket.on).toHaveBeenCalledWith("message", expect.any(Function));
  });

  it('should start watering when the start button is clicked', () => {
  });

  it('should stop watering when the stop button is clicked', () => {
  });
});