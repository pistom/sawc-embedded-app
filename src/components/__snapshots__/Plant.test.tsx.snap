// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Plant></Plant> should render the component 1`] = `
<div
  class="w-full md:w-1/2 lg:w-1/3"
>
  <div
    class="plant relative"
  >
    <span
      class="absolute bottom-2 right-2 text-sm text-slate-400 "
    >
      Not synced 
      <svg
        aria-hidden="true"
        class="align-sub h-5 w-5 inline-block"
        data-slot="icon"
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18 1.5c2.9 0 5.25 2.35 5.25 5.25v3.75a.75.75 0 0 1-1.5 0V6.75a3.75 3.75 0 1 0-7.5 0v3a3 3 0 0 1 3 3v6.75a3 3 0 0 1-3 3H3.75a3 3 0 0 1-3-3v-6.75a3 3 0 0 1 3-3h9v-3c0-2.9 2.35-5.25 5.25-5.25Z"
        />
      </svg>
    </span>
    <img
      alt=""
      class="image bg-slate-200"
      src="/plants/test"
    />
    <div
      class="details"
    >
      <a
        class="editBtn"
        href="/output/edit/device/output"
      >
        <svg
          aria-hidden="true"
          class="h-5 w-5 text-slate-300 absolute top-2 right-2"
          data-slot="icon"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M11.078 2.25c-.917 0-1.699.663-1.85 1.567L9.05 4.889c-.02.12-.115.26-.297.348a7.493 7.493 0 0 0-.986.57c-.166.115-.334.126-.45.083L6.3 5.508a1.875 1.875 0 0 0-2.282.819l-.922 1.597a1.875 1.875 0 0 0 .432 2.385l.84.692c.095.078.17.229.154.43a7.598 7.598 0 0 0 0 1.139c.015.2-.059.352-.153.43l-.841.692a1.875 1.875 0 0 0-.432 2.385l.922 1.597a1.875 1.875 0 0 0 2.282.818l1.019-.382c.115-.043.283-.031.45.082.312.214.641.405.985.57.182.088.277.228.297.35l.178 1.071c.151.904.933 1.567 1.85 1.567h1.844c.916 0 1.699-.663 1.85-1.567l.178-1.072c.02-.12.114-.26.297-.349.344-.165.673-.356.985-.57.167-.114.335-.125.45-.082l1.02.382a1.875 1.875 0 0 0 2.28-.819l.923-1.597a1.875 1.875 0 0 0-.432-2.385l-.84-.692c-.095-.078-.17-.229-.154-.43a7.614 7.614 0 0 0 0-1.139c-.016-.2.059-.352.153-.43l.84-.692c.708-.582.891-1.59.433-2.385l-.922-1.597a1.875 1.875 0 0 0-2.282-.818l-1.02.382c-.114.043-.282.031-.449-.083a7.49 7.49 0 0 0-.985-.57c-.183-.087-.277-.227-.297-.348l-.179-1.072a1.875 1.875 0 0 0-1.85-1.567h-1.843ZM12 15.75a3.75 3.75 0 1 0 0-7.5 3.75 3.75 0 0 0 0 7.5Z"
            fill-rule="evenodd"
          />
        </svg>
      </a>
      <h5
        class="title"
      >
        test
      </h5>
      <div
        class="form flex gap-3 mb-3 font-normal text-gray-700"
      >
        <div
          class="flex items-center"
        >
          <input
            class="unit"
            type="number"
            value="5"
          />
          <span
            class="unit-label"
          >
            ml
          </span>
        </div>
        <button
          class="waterBtn btn bg-emerald-700"
        >
          Water
        </button>
      </div>
    </div>
  </div>
</div>
`;
